proof_of_stake/src/lib.rs

```diff
@@ -0,0 +1,1814 @@
+            self.init_staking_reward_account(
+                &staking_reward_address,
+                &staking_reward_key,
+            );
+        }
+        self.write_validator_set(&validator_set);
+        self.write_total_voting_power(&total_voting_power);
+        // Credit the bonded tokens to the PoS account
+        self.credit_tokens(
+            &Self::staking_token_address(),
+            &Self::POS_ADDRESS,
+            total_bonded_balance,
+        );
+        Ok(())
+    }
+
+    /// Calls a closure on each validator update element.
+    fn validator_set_update(
+        &self,
+        current_epoch: impl Into<Epoch>,
+        f: impl FnMut(ValidatorSetUpdate<Self::PublicKey>),
+    ) {
+        let current_epoch = current_epoch.into();
+        let validators = self.read_validator_set();
+        let validators = validators.get(current_epoch).unwrap();
+        let active_validators = validators.active.iter().map(
+            |WeightedValidator {
+                 voting_power,
+                 address,
+             }: &WeightedValidator<Self::Address>| {
+                let consensus_key = self
+                    .read_validator_consensus_key(address)
+                    .unwrap()
+                    .get(current_epoch)
+                    .unwrap()
+                    .clone();
+                ValidatorSetUpdate::Active(ActiveValidator {
+                    consensus_key,
+                    voting_power: *voting_power,
+                })
+            },
+        );
+        // TODO we only need inactive validators that were active in the last
+        // update
+        let inactive_validators = validators.inactive.iter().map(
+            |WeightedValidator {
+                 voting_power: _,
+                 address,
+             }: &WeightedValidator<Self::Address>| {
+                let consensus_key = self
+                    .read_validator_consensus_key(address)
+                    .unwrap()
+                    .get(current_epoch)
+                    .unwrap()
+                    .clone();
+                ValidatorSetUpdate::Deactivated(consensus_key)
+            },
+        );
+        active_validators.chain(inactive_validators).for_each(f)
+    }
+
+    /// Apply a slash to a byzantine validator for the given evidence.
+    fn slash(
+        &mut self,
+        params: &PosParams,
+        current_epoch: impl Into<Epoch>,
+        evidence_epoch: impl Into<Epoch>,
+        evidence_block_height: impl Into<u64>,
+        slash_type: SlashType,
+        validator: &Self::Address,
+    ) -> Result<(), SlashError<Self::Address>> {
+        let current_epoch = current_epoch.into();
+        let evidence_epoch = evidence_epoch.into();
+        let rate = match &slash_type {
+            SlashType::DuplicateVote => params.duplicate_vote_slash_rate,
+            SlashType::LightClientAttack => {
+                params.light_client_attack_slash_rate
+            }
+        };
```

Comment:
I anticipate we will want to add more SlashTypes, e.g. from the DKG protocol. To keep from maintaining this match statement, maybe we can implement a `fn get_slash_rate(&self, params: &PosParams) -> u64` on the `SlashType` enum instead

Reply:
Yup, good point. I've added it in c464d1541d99120bfda0ce9d7865c729191df025